ruby/ql/src/queries/security/cwe-312/CleartextStorage.ql

```diff
@@ -0,0 +1,24 @@
+/**
+ * @name Clear-text storage of sensitive information
+ * @description Storing sensitive information without encryption or hashing can
+ *              expose it to an attacker.
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 7.5
+ * @precision high
+ * @id rb/clear-text-storage-sensitive-data
+ * @tags security
+ *       external/cwe/cwe-312
+ *       external/cwe/cwe-359
+ *       external/cwe/cwe-532
+ */
+
+import ruby
+import codeql.ruby.security.CleartextStorageQuery
+import codeql.ruby.DataFlow
+import DataFlow::PathGraph
+
+from Configuration config, DataFlow::PathNode source, DataFlow::PathNode sink
+where config.hasFlowPath(source, sink)
+select sink.getNode(), source, sink, "Sensitive data returned by $@ is stored here.",
+  source.getNode(), source.getNode().(Source).describe()
```

Comment:
Looking at some of the results in the DCA evaluation, I wonder if we should report these alerts at the source location rather than the sink location. There seem to be a lot of cases in `diaspora/diaspora` where a password value flows from a test to a call to `User.create` or similar. I worry that, since the alert is located at the sink, in `models/user.rb`, any code-scanning filtering to remove alerts in test files will let it pass through.

Reply:
This seems like a good idea overall. The case of a sink in test code and source in production code seems far less likely to occur in practice.